src/backend/distributed/commands/sequence.c

```diff
@@ -213,6 +213,46 @@ ExtractDefaultColumnsAndOwnedSequences(Oid relationId, List **columnNameList,
 }
 
 
+/*
+ * ColumnDefaultsToNextVal returns true if the column with attrNumber
+ * has a default expression that contains nextval().
+ */
+bool
+ColumnDefaultsToNextVal(Oid relationId, AttrNumber attrNumber)
+{
+	Relation relation = RelationIdGetRelation(relationId);
+
+	TupleDesc relTupDesc = RelationGetDescr(relation);
+	TupleConstr *relTupConstraints = relTupDesc->constr;
+	if (relTupConstraints == NULL)
+	{
+		/* none of the columns have a DEFAULT expression */
+		RelationClose(relation);
+		return false;
+	}
+
+	for (uint16 defExprIndex = 0; defExprIndex < relTupConstraints->num_defval;
+		 defExprIndex++)
+	{
+		AttrDefault *defValue = &(relTupConstraints->defval[defExprIndex]);
+		if (defValue->adnum != attrNumber)
+		{
+			continue;
+		}
+
+		Node *defExpr = (Node *) stringToNode(defValue->adbin);
+
+		RelationClose(relation);
+
+		return contain_nextval_expression_walker(defExpr, NULL);
+	}
+
+	/* column doesn't have a DEFAULT expression */
```

Comment:
what if the default is a volatile function? `x int DEFAULT volatile_f()`?

I'm wondering if we should expand next_val check to volatile (or probably non-immutable) check including #6340?

Because we cannot afford to have the default on the shard and the default on the table produce any different values in any case.

what do you think?

Reply:
Hmm, I'm not sure what makes allowing such expressions on shards "more dangerous" than allowing them on plain postgres tables.

In other words, I couldn't see any reasons why to disallow them while postgres doesn't have such a limitation.

But in any case, to me this sounds like a broader discussion than this PR's context.

Reply:
> I cannot think of a case where this would cause a problem right now, but I am also not sure that this is not going to cause any issues :)

Thought a bit more about that and couldn't see anything that could break DEFAULT value assignment on shards, if I'm not missing something obvious :)